Fixed #28107 -- Added DatabaseFeatures.allows_group_by_selected_pks_on_model() to allow enabling optimization for unmanaged models.

Tasssadar opened the pull request.

felixxm commented on django/db/backends/postgresql/features.py:
```diff
@@ -67,3 +67,24 @@ def is_postgresql_10(self):
+            return True
+
+        try:
+            return not model._meta._is_view
```
If we want to add a new `_is_view` option to `Meta` it should be documented, but I don't think it is necessary.

Tasssadar replied:
The value is used as a cache for the query so that it isn't done for every access, see line 89. I can switch to the introspecition + lru_cache approach if that's better, but I'm not sure how to mock that value in tests.

felixxm replied:
OK, makes sense. Name is quite misleading because we're checking if it is a table, please rename `_is_view` to `_is_table`. Also, IMO we still should add `_is_table()` (without caching) hook to the introspection.

felixxm commented on django/db/backends/base/features.py:
```diff
@@ -308,3 +308,9 @@ def supports_transactions(self):
+        """ When allows_group_by_selected_pks is True, this method is used to
+        determine whether the particular model supports it.
+        Default: return model._meta.managed"""
+        return model._meta.managed
```
This optimization is used only for backends with `allows_group_by_selected_pks` feature, we should use it here, e.g.
```python
def allows_group_by_selected_pks_on_model(self, model):
    if self.allows_group_by_selected_pks:
        return model._meta.managed
    return False
```

Tasssadar replied:
That is handled in the compiler.py, this method is never called unless `allows_group_by_selected_pks` is True.

felixxm replied:
I know but what it will use it somewhere else?

felixxm replied:
We cannot return `True` if a backend doesn't support it at all.

carltongibson replied:
Just _unresolved_ this to keep it visible. Reading the code as it is now, I was "the `allows_group_by_selected_pks` is redundant — this method is never called in that case". 

If it's a pre-condition/guard, I'd usually separate it make obviously so:

```suggestion
def allows_group_by_selected_pks_on_model(self, model):
    if not self.allows_group_by_selected_pks:
        return False
    # Not a real comment: ...with the guard out of the way, actual method...
    return model._meta.managed
```

...but it's fine as it is.

felixxm replied:
Agreed, fixed.

felixxm commented on django/db/backends/base/features.py:
```diff
@@ -308,3 +308,9 @@ def supports_transactions(self):
+    def allows_group_by_selected_pks_on_model(self, model):
+        """ When allows_group_by_selected_pks is True, this method is used to
+        determine whether the particular model supports it.
+        Default: return model._meta.managed"""
```
I think we can remove this docstring.

felixxm commented on django/db/backends/postgresql/features.py:
```diff
@@ -67,3 +67,24 @@ def is_postgresql_10(self):
+                        AND c.relkind = 'r'
+                        AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
+                        AND pg_catalog.pg_table_is_visible(c.oid)
+                     """, (model._meta.db_table,))
```
I think we should move this hook to the PostgreSQL's introspection e.g. `is_table(table_name)`:
```python
def is_table(self, table_name):
    with self.connection.cursor() as cursor:
        cursor.execute("""
            SELECT TRUE
            FROM pg_class c
            JOIN pg_namespace n ON c.relnamespace = n.oid
            WHERE
               c.relname = %s AND
               c.relkind = 'r' AND
               n.nspname NOT IN ('pg_catalog', 'pg_toast') AND
               pg_catalog.pg_table_is_visible(c.oid)
       """, [model._meta.db_table])
       return cursor.fetchone() is not None
```
Also this should be cached e.g. `lru_cache(maxsize=512)` we don't want to perform an extra query each time.

felixxm commented on django/db/backends/postgresql/features.py:
```diff
@@ -74,17 +76,9 @@ def allows_group_by_selected_pks_on_model(self, model):
-                     """, (model._meta.db_table,))
-                model._meta._is_view = cursor.fetchone() is None
-            return not model._meta._is_view
+                model._meta._is_ordinary_table = intr.is_ordinary_table(cursor, model._meta.db_table)
```
You can use `self.connection.introspection.is_ordinary_table(...)`.

Tasssadar replied:
Didn't know that, thanks! I also see the introspection is a child of `BaseDatabaseIntrospection`, is it okay that the Posgres' introspection has the extra `is_ordinary_table` method that's not defined in `BaseDatabaseIntrospection`?

felixxm replied:
Yes, that's fine.

felixxm commented on django/db/backends/postgresql/introspection.py:
```diff
@@ -220,3 +220,18 @@ def get_constraints(self, cursor, table_name):
+    def is_ordinary_table(self, cursor, table_name):
+        """
+        Returns true if the table is an ordinary table, as opposed to a view.
+        """
```
I don't see much value in this docstring.

felixxm commented on tests/aggregation_regress/tests.py:
```diff
@@ -1318,19 +1318,43 @@ def assertQuerysetResults(queryset):
+            self.assertEqual(len(grouping), 1 if connection.features.allows_group_by_pk else 2)
+            self.assertIn('id', grouping[0][0])
+            self.assertNotIn('name', grouping[0][0])
+            self.assertNotIn('contact', grouping[0][0])
```
`grouping[0][0]` is a name of the first column, so these two assertions are unnecessary:
```python
self.assertNotIn('name', grouping[0][0])
self.assertNotIn('contact', grouping[0][0]) 
```

felixxm commented on tests/aggregation_regress/tests.py:
```diff
@@ -1318,19 +1318,43 @@ def assertQuerysetResults(queryset):
+            self.assertEqual(len(grouping), 1 if connection.features.allows_group_by_pk else 2)
+            self.assertIn('id', grouping[0][0])
+            self.assertNotIn('name', grouping[0][0])
+            self.assertNotIn('contact', grouping[0][0])
```
`grouping[0][0]` is a name of the first column, so these two assertions are unnecessary:
```python
self.assertNotIn('name', grouping[0][0])
self.assertNotIn('contact', grouping[0][0]) 
```

felixxm commented on tests/aggregation_regress/tests.py:
```diff
@@ -1318,19 +1318,43 @@ def assertQuerysetResults(queryset):
+            self.assertIn('id', grouping[0][0])
+            self.assertNotIn('name', grouping[0][0])
+            self.assertNotIn('contact', grouping[0][0])
+            self.assertEqual(Author._meta._is_ordinary_table, True)
```
`assertEqual` -> `assertTrue`

felixxm commented on tests/aggregation_regress/tests.py:
```diff
@@ -1318,19 +1318,43 @@ def assertQuerysetResults(queryset):
+            self.assertIn('id', grouping[0][0])
+            self.assertNotIn('name', grouping[0][0])
+            self.assertNotIn('contact', grouping[0][0])
+            self.assertEqual(Book._meta._is_ordinary_table, True)
```
`assertEqual` -> `assertTrue`

Tasssadar commented on django/db/backends/postgresql/features.py:
```diff
@@ -67,3 +69,16 @@ def is_postgresql_10(self):
     has_phraseto_tsquery = property(operator.attrgetter('is_postgresql_9_6'))
     supports_table_partitions = property(operator.attrgetter('is_postgresql_10'))
+
+    # Views do not support the allows_group_by_selected_pks optimization"
```
The reason for this comment was to highlight the reason why is this distinction necessary (PosgreSQL only supports this on ordinary tables). How about:

```python
    # PostgreSQL only supports this optimization on ordinary tables.
    def allows_group_by_selected_pks_on_model(self, model):
        # Assume that managed models are ordinary tables.
        if model._meta.managed:
            return True

        try:
            return model._meta._is_ordinary_table
        except AttributeError:
            ...
```

Tasssadar commented:
Thanks for your comments!

>     1. The current code is not schema aware and could break with usage of `SET search_path` and friends or namedspaced `db_table` (e.g. `Meta.db_table = '"schema"."table"`.

I copied the WHERE partly from `get_table_description` in PostgreSQL's introspection.py, it didn't seem to be aware of schemas either so I didn't bother :/

>     2. We should try to find a more appropriate name than _ordinary_ to categorize such tables.

That's the nomenclature used by Postgres docs, https://www.postgresql.org/docs/9.3/catalog-pg-class.html , that's why I put it everywhere (instead of just `is_table`).

>     3. Using `Model._meta` as a cache store is a big no-no IMO because it doesn't take connection aliases into account. Results should be cached using a `(alias, schema, table_name)` namespace.
> 
> Given the rare cases when this is useful and the complexity of implementing a bullet proof caching mechanism I'd favor simply exposing the `allows_group_by_selected_pks_on_model` hook and letting developers override it they deem it necessary.

Where should this hook be? I can't see a way to override connection features from the user app without creating a new DB engine :/

charettes commented:
Hey @Tasssadar 

> I copied the WHERE partly from get_table_description in PostgreSQL's introspection.py, it didn't seem to be aware of schemas either so I didn't bother :/

`get_table_description` doesn't implement any form of caching so it doesn't have to deal with this extra layer of complexity.

> That's the nomenclature used by Postgres docs

TIL, I didn't know that was the name used by Postgres. Thanks for pointing that out. Reading the documentation makes me wonder if the optimization works for foreign and partitioned tables though. I guess that's another reason why this should be left to the user.

> Where should this hook be? I can't see a way to override connection features from the user app without creating a new DB engine :/

That would effectively require the user to subclass the Postgres db engine which is not as daunting as it might seems.

Tasssadar commented:
I have removed the autodetection and just added a hook that just returns `managed` by default. I'm not sure about the tests, I think it's not necessary to change them for this hook and keep them as is.

I tried writing the subclassed engine, and it's not bad:
```python
from django.db.backends.postgresql import base, features

class DatabaseFeatures(features.DatabaseFeatures):
    def allows_group_by_selected_pks_on_model(self, model):
        return True

class DatabaseWrapper(base.DatabaseWrapper):
    features_class = DatabaseFeatures
```

I'm not sure how you're merging PRs, there's a lot of commits here now with useless code, I can rebase it to master if you with.

Tasssadar commented:
Hello, any followup please? The tests failed, but I'm fairly sure it was something in the test environment wrong, not the code.

felixxm commented:
@Tasssadar Thanks. I rebased from master, squashed commits, restored one test and added release notes.

carltongibson commented on docs/releases/3.0.txt:
```diff
@@ -312,6 +312,13 @@ Models
+  to ``GROUP BY`` only by selected models' primary keys for unmanaged models on
+  PostgreSQL, which is disabled by default. If you want to use this feature you
+  have to subclass the PostgreSQL database engine and override
+  ``allows_group_by_selected_pks_on_model()``.
```
I think this sentence isn't enough. 

* You have to override the method on the features class, 
* Then, override the database wrapper, 
* Then, save it in the right plave, i.e. a `my_backend.base`, pointing `DATABASES` (`BACKEND`) to `my_backend`. 

I think we need an example of that. 

SO here, broken up for consideration: 

```
The new ``allows_group_by_selected_pks_on_model()`` method of
``django.db.backends.base.BaseDatabaseFeatures`` allows optimization of ``GROUP
BY`` clauses to require only the selected models' primary keys.

This optimization is currently only supported on PostgreSQL. By default,
it is only applied to **managed models**.

To enable the the GROUP BY primary key-only optimization for unmanaged models,
subclass the PostgreSQL database engine, overriding the features class
``allows_group_by_selected_pks_on_model()`` as you require. See LINK TO FULL
EXAMPLE for an example.
```

Then in FULL EXAMPLE we need the full example @Tasssadar came up with [in this comment](https://github.com/django/django/pull/11606#issuecomment-517255305), plus the file to save it in, and the `DATABASES` setting example. 

Open to suggestions as to where but I would create a new section in `databases.txt` called, `Subclassing the built-in backends`, just before `Using a 3rd-party database backend`, to hold FULL EXAMPLE (and be the target of LINK TO FULL EXAMPLE).

carltongibson replied:
@Tasssadar: Do you this you have bandwidth to make these docs changes? (Thanks if you do! 🎉)

Tasssadar replied:
Back from vacation, I tried to write an example for the docs, please take a look :)

Tasssadar commented:
Rebased due to conflicts, the failing test is caused by ed2d411aa84efc76baba3adf0d0f99df0e44ba57 and not this PR.

felixxm commented:
@Tasssadar Yes my mistake, I will fix this in a minute.

felixxm commented:
@Tasssadar @carltongibson I rebased from master, moved _"Subclassing the built-in database backends"_ to a separate commit and rephrased it a bit. IMO we're ready to merge :confetti_ball:

felixxm commented:
@Tasssadar Thanks, welcome aboard :boat: 

@carltongibson Thanks for checking :+1:

Tasssadar commented:
Thank you! This first release with this included will be 3.0 in December, right?

felixxm commented:
Yes.
